Update to rust-url 1.0

SimonSapin opened the pull request:
**Do not merge yet:** rust-url 1.0 is not published yet and may still get breaking changes. The goal of this PR for now is to demonstrate API usage.

Depends on:
- <s>https://github.com/servo/rust-url/pull/176</s>
- <s>https://github.com/alexcrichton/cookie-rs/pull/42</s>
- <s>https://github.com/hyperium/hyper/pull/740</s>
- <s>https://github.com/cyderize/rust-websocket/pull/70</s>
- <s>https://github.com/jgraham/webdriver-rust/pull/28</s>

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/9840)

<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/selector_matching.rs, components/style/parser.rs, components/style/font_face.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

asajeffrey commented:
Removing the `url!` plugin means there's now urls being parsed at run-time rather than compile-time (e.g. `about:blank`). This is a step back in performance, and introduces some `unwrap`s that weren't there before. The performance hit is probably pretty negligible, and the `unwrap`s are annoying but most likely safe.

At some point, rust might get constant functions, then we could define (e.g) `const ABOUT_BLANK: Url = Url::parse("about:blank").unwrap();`.

asajeffrey commented:
Reviewed 63 of 63 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[components/devtools/lib.rs, line 278 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsHhJI4yfoiZ5PkIWN:-KBsHhJJPQHIaTyqM6Xs:2089077573) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/devtools/lib.rs#L278)):*
It would be nice to be able to convert directly from a Url to String. I think this is unnecessarily cloning the URL string representation.

---

*[components/layout/layout_thread.rs, line 966 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsJ50Qhn9opAlIpLw-:-KBsJ50Qhn9opAlIpLw0:-2058511884) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/layout/layout_thread.rs#L966)):*
Since URL implements Debug shouldn't this be `debug!("layout: received layout request for: {:?}", self.url);`?

---

*[components/net/hsts.rs, line 126 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsLWsagHgXjfI4s1kb:-KBsLWsbySTK2i6fbOmz:1386554474) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/net/hsts.rs#L126)):*
Slightly unintuitive having set_scheme return a Result which needs to be unwrapped. Would it be worth exporting a Scheme type containing pre-parsed schemes, then having set_scheme just return ()? Or is this because of the error conditition "This URL is non-relative and the new scheme is one of `http`, `https`, `ws`, `wss`, `ftp`, or `gopher`" (whcih AFAICT is unenforced)?

---

*[components/net/http_loader.rs, line 691 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsMw-cPzTXfzh6Dx4O:-KBsMw-cPzTXfzh6Dx4P:-1222243585) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/net/http_loader.rs#L691)):*
Can we do better than reparsing the URL? Can rust-url export a helper method for this? Not sure about the `unwrap()` here.

---

*[components/net/storage_thread.rs, line 219 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsNJy3pAXkb1-gdNaX:-KBsNJy3pAXkb1-gdNaY:264720186) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/net/storage_thread.rs#L219)):*
Much nicer!

---

*[components/net/fetch/methods.rs, line 541 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsNtXi5gaJYNCERpPc:-KBsNtXj20bS7rzsH8b7:-1785381943) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/net/fetch/methods.rs#L541)):*
Bikeshed the name `join`?

---

*[components/net_traits/hosts.rs, line 74 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsOH3ClFgb8-iTLIBk:-KBsOH3DhI7QqyeaT3AX:77416319) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/net_traits/hosts.rs#L74)):*
Should url export a helper function `set_ip_host` for this?

---

*[components/script/cors.rs, line 429 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsOak952AWbYOO3DeP:-KBsOak952AWbYOO3DeQ:1260479927) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/script/cors.rs#L429)):*
Inconsistent line break.

---

*[components/script/script_thread.rs, line 1864 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsOx8xDHJufhdwuT2y:-KBsOx8xDHJufhdwuT2z:-1504287282) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/script/script_thread.rs#L1864)):*
What happened to steps 5-8? Are they now in `lossy_utf8_percent_decode`? If so, a comment would be good.

---

*[components/script/dom/element.rs, line 1081 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsPLd9Hq_X5CAlV9Pa:-KBsPLd9Hq_X5CAlV9Pb:812884083) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/script/dom/element.rs#L1081)):*
I think this unnecessarily clones the url.

---

*[components/script/dom/macros.rs, line 102 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsPdhUeuYh79qdR0ap:-KBsPdhUeuYh79qdR0aq:-622211434) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/script/dom/macros.rs#L102)):*
Is this cloning the URL?

---

*[components/script/dom/urlhelper.rs, line 12 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsPviZsFo4pH18Uy2Y:-KBsPviZsFo4pH18Uy2Z:-566878122) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/script/dom/urlhelper.rs#L12)):*
OK, this is much nicer!

---

*[tests/unit/util/opts.rs, line 18 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsQgxlPbRXVxfcAswN:-KBsQgxlPbRXVxfcAswO:-1047479894) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/tests/unit/util/opts.rs#L18)):*
Not obvious why path_segments() needs an unwrap(). Should it just return the empty iterator if there are none?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9840)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
There's only a handful of constant URLs. Put them in the URL crate?

bors-servo commented:
:umbrella: The latest upstream changes (presumably #9715) made this pull request unmergeable. Please resolve the merge conflicts.

asajeffrey commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9840)*
<!-- Sent from Reviewable.io -->

SimonSapin commented:
@frewsxcv said on IRC that he mostly implemented `url!` because it seemed easy and to learn Rust compiler plugin. The performance benefit was never measured, but I’d bet it’s negligible in Servo.

Re constant URLs, Rust’s `const fn` is currently very limited and can definitely not run at compile-time an algorithm like `Url::parse`.

---

Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[components/devtools/lib.rs, line 278 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsHhJI4yfoiZ5PkIWN:-KDZgixzce1bywlc76aW:-427166786) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/devtools/lib.rs#L278)):*
I’ve added `Url::into_string` and used it here.

---

*[components/layout/layout_thread.rs, line 966 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsJ50Qhn9opAlIpLw-:-KDYotIWA0ajnUNmwjaY:833803563) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/layout/layout_thread.rs#L966)):*
Now that I check I see that `RefCell` implements `Debug` so that would work. But it’s [more verbose](https://github.com/rust-lang/rust/blob/d6af19b89c477708ed0f5341ab7a26f1764f1ffa/src/libcore/fmt/mod.rs#L1599-L1611), printing something like `Refcell { value: "http://example.net" }`

---

*[components/net/hsts.rs, line 126 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsLWsagHgXjfI4s1kb:-KDYpgJQjZwHPso4MbOe:1578112443) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/net/hsts.rs#L126)):*
It’s not just `set_scheme`, some other setter methods return `Result<(), ()>` to indicate that they may fail (and do nothing) in some cases. As you mention for scheme there are parsing errors, but even for just that I’m not convinced that a pre-parsed `Scheme` type is more convenient than slapping `.unwrap()` here and there. There is an other case that I wrote up in the doc-comment and, you’re right, forgot to implement. I’ve just done that.

---

*[components/net/http_loader.rs, line 691 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsMw-cPzTXfzh6Dx4O:-KDZXutHWXRzWOX6k-TV:-955551053) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/net/http_loader.rs#L691)):*
With an input of e.g. `view-source:https://google.com?q=foo`, there is no host and the path is `https://google.com`. This is completely different from what we want, so yes re-parsing is necessary.

`unwrap()` here can definitely panic, but it already could before.

Both issues could be solved by changing the pattern from `view-source:{absolute url}` to `view-source+{absolute url}`. That way we can check for `url.scheme().startswith("view-source+")` and extract the "inner" URL with `url.set_scheme(&scheme["view-source+".len()..])`. However that would be a functional change, so I’d prefer to do it in a separate PR.

---

*[components/net/fetch/methods.rs, line 541 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsNtXi5gaJYNCERpPc:-KDZZX0XEgdAf2gro-7u:139608400) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/net/fetch/methods.rs#L541)):*
There is precedent in `std::path::Path` with its `join` method and the `urlparse.urljoin` function in Python. And this method already exists in 0.5 https://github.com/servo/rust-url/pull/153

Do you have another suggestion?

---

*[components/net_traits/hosts.rs, line 74 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsOH3ClFgb8-iTLIBk:-KDZ_OGCwynYtUSSk-0g:-876204362) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/net_traits/hosts.rs#L74)):*
Done. I had removed it previously since `std::net::IpAddr` was not yet `#[stable]` in the stable channel, but it is now with Rust 1.7.

---

*[components/script/cors.rs, line 429 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsOak952AWbYOO3DeP:-KDZ_ibD9NQcU8eWVaIV:1758359711) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/script/cors.rs#L429)):*
Fixed.

---

*[components/script/script_thread.rs, line 1864 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsOx8xDHJufhdwuT2y:-KDZeGtWR1hOFmiqAMlJ:1760560041) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/script/script_thread.rs#L1864)):*
This `let` statement does everything in 5 to 8. I’ve introduced an intermediate variable and rephrased the comment.

---

*[components/script/dom/element.rs, line 1081 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsPLd9Hq_X5CAlV9Pa:-KDZhFpb1Xc191lRDb7E:1758359711) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/script/dom/element.rs#L1081)):*
Fixed.

---

*[components/script/dom/macros.rs, line 102 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsPdhUeuYh79qdR0ap:-KDZhg48B2zp7_FaU9iE:1758359711) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/components/script/dom/macros.rs#L102)):*
Fixed.

---

*[tests/unit/util/opts.rs, line 18 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9840#-KBsQgxlPbRXVxfcAswN:-KDZhnE9wUm5a8aGzpLp:1637678682) ([raw file](https://github.com/servo/servo/blob/50915cd77be7ff5c7540c946b4362e009c8c4f12/tests/unit/util/opts.rs#L18)):*
Non-relative URLs (soon to be called "URLs that cannot be a base" https://github.com/whatwg/url/issues/89) don’t have a path made of slash-separated segments, they only have a single unstructured string. This is different from a path made of zero segments.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9840#-:-KDZiChSiQDOQORJzzJb:1817605842)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #9887) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #10188) made this pull request unmergeable. Please resolve the merge conflicts.

asajeffrey commented:
@SimonSapin what's the status on this PR?

SimonSapin commented:
@asajeffrey This needs to happen before it can be merged:

* I write tests for setters in rust-url (I’ve made a test harness for this today) and fix corresponding bugs (valenting has already found some).
* Ideally https://github.com/servo/rust-url/pull/176 should be fully reviewed, but I understand it’s quite daunting at this point. Lesson learned: for big rewrites in the future, avoid mega-PRs and rather have smaller PRs reviewed along the way into a feature branch.
* https://github.com/servo/rust-url/pull/176 is merged and published on crates.io
* Similarly for https://github.com/alexcrichton/cookie-rs/pull/42, https://github.com/hyperium/hyper/pull/740, and https://github.com/cyderize/rust-websocket/pull/70.
* Update version numbers in `Cargo.toml` files in this PR.

SimonSapin committed:
Upgrade to rust-url 1.0 and hyper 0.9

SimonSapin commented:
@bors-servo try

(Using `[replace]` in `Cargo.toml` for dependencies not updated/published yet)

bors-servo commented:
:hourglass: Trying commit ff5368c with merge f38d81c...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/431)

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 24d9d06 with merge 8b1ad8a...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/490)

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 9621b62 with merge bcc4379...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/416)

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit adb590e with merge e28e6e8...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/492)

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 36a0381 with merge bc3758b...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/477)

SimonSapin commented:
```
  ▶ OK [expected CRASH] /webgl/conformance-1.0.3/conformance/more/conformance/quickCheckAPI-S_V.html

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/more/conformance/quickCheckAPI-S_V.html:
  │ FAIL [expected PASS] WebGL test #0: testValidArgs
  │   → assert_true: testValidArgs expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:900:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1397:20
  │ test@http://web-platform.test:8000/resources/testharness.js:495:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:899:1
  │ runTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:188:5
  │ initTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:922:5
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:947:7

  ▶ Unexpected subtest result in /websockets/constructor/018.html:
  └ PASS [expected FAIL] WebSockets: NULL char in url
```

The first two are intermittent (#10698). The third is related to URL parsing, I’ve now removed `expected: FAIL`.

SimonSapin committed:
Correctly initialize URL.searchParams

SimonSapin committed:
Reset searchParam in URL.href setter

https://github.com/whatwg/url/issues/117

SimonSapin committed:
Only test expected origin when it is present

https://github.com/w3c/web-platform-tests/pull/2584

SimonSapin commented:
I think this is now ready to land (after review).

SimonSapin commented:
Reviewable is showing unrelated changes, probably due to massive and repeated rebase. It may be easier to look at the github diff instead. Maybe.

asajeffrey commented:
I see what you mean about reviewable. Some minor nits.

---

Reviewed 36 of 53 files at r6, 5 of 11 files at r8, 2 of 6 files at r11, 28 of 30 files at r12, 5 of 5 files at r13.
Review status: 75 of 77 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[components/net/chrome_loader.rs, line 22 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyFGQyGkzKFrsBzSAd:-KFyFGQyGkzKFrsBzSAe:222514300) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/net/chrome_loader.rs#L22)):*
Do we not need to normalize the path to stop things like `chrome:../../../p0wned.txt`? A test might be nice.

---

*[components/script/script_thread.rs, line 1565 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyVtNlj_jO-GMyRDG2:-KFyVtNlj_jO-GMyRDG3:747765772) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/script/script_thread.rs#L1565)):*
Do you mean 4-6?

---

*[components/script/script_thread.rs, line 1566 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyVzswrHh0iZEHRWI0:-KFyVzswrHh0iZEHRWI1:436021528) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/script/script_thread.rs#L1566)):*
4. is actually "Append the first string of urlRecord's path component to script source." Doesn't this mean that if the original URL is `http://example.com/foo/bar/baz?stuff` then the result should be `foo?stuff`?

---

*[components/script/dom/htmlanchorelement.rs, line 66 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyYTi2Ok67ByDaDtJh:-KFyYTi2Ok67ByDaDtJi:264720186) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/script/dom/htmlanchorelement.rs#L66)):*
Much nicer!

---

*[components/script/dom/htmlbaseelement.rs, line 87 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyYj0Z8PIEcChaJy-6:-KFyYj0Z8PIEcChaJy-7:-145599035) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/script/dom/htmlbaseelement.rs#L87)):*
Easier to read as `DOMString::from(url_record.map(URL::as_str).unwrap_or(""))`?

---

*[components/script/dom/urlhelper.rs, line 7 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFy_aEn6wfoiiXjLVSo:-KFy_aEpMmnMzlVKp8wo:-1124678419) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/script/dom/urlhelper.rs#L7)):*
I don't understand why these are in `quirks`, they don't look quirky to me!

---

*[components/script/dom/urlsearchparams.rs, line 34 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFy_ws1cSIud1lrNRCl:-KFy_ws1cSIud1lrNRCm:-754146343) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/script/dom/urlsearchparams.rs#L34)):*
Isn't this `map_or_default`?

---

*[resources/badcert.html, line 6 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyc35Y9wkvoKDwLBas:-KFyc35Y9wkvoKDwLBat:480739907) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/resources/badcert.html#L6)):*
OK, I don't understand this change. Shouldn't `chrome` URLs be of the form `chrome://...` (e.g. https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XUL/Tutorial/The_Chrome_URL)?

---

*[tests/unit/net/chrome_loader.rs, line 10 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFycZi0JFtsCMwEorAO:-KFycZi0JFtsCMwEorAP:1066307092) ([raw file](https://github.com/servo/servo/blob/0e0e902edd200359db6ae06cd149fb1b0a97fcc6/tests/unit/net/chrome_loader.rs#L10)):*
Again, I don't understand the change to chrome URLs.

---

*[tests/wpt/metadata/url/url-constructor.html.ini, line 5 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFydTnY1JqV6HQnOxea:-KFydTnZzG4SWwdWIz9k:-372022388) ([raw file](https://github.com/servo/servo/blob/0e0e902edd200359db6ae06cd149fb1b0a97fcc6/tests/wpt/metadata/url/url-constructor.html.ini#L5)):*
Good to see a lot of `FAIL`s disappearing!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9840#-:-KFyeSo1yLYsHZemfWMX:1610322203)*
<!-- Sent from Reviewable.io -->

SimonSapin commented:
Review status: 75 of 77 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[components/net/chrome_loader.rs, line 22 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyFGQyGkzKFrsBzSAd:-KFyfWUhRmIjF8QQBzdg:-175793448) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/net/chrome_loader.rs#L22)):*
`..` path segments are resolved in the URL parser: https://github.com/servo/rust-url/blob/d5a84c9/src/parser.rs#L815

`tests/unit/net/chrome_loader.rs` has some tests with `..` path segments.

---

*[components/script/script_thread.rs, line 1565 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyVtNlj_jO-GMyRDG2:-KFyhPHWvejF3YVmcqpH:-1124956784) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/script/script_thread.rs#L1565)):*
Hum, maybe the spec has changed since that code was first written?

---

*[components/script/script_thread.rs, line 1566 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyVzswrHh0iZEHRWI0:-KFyihDvAeSbmVLrPfgf:1957670570) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/script/script_thread.rs#L1566)):*
The URL spec says "A URL [record]’s path is a list of zero or more ASCII string holding data, usually identifying a location in hierarchical form. It is initially the empty list. ", but for cannot-be-a-base URLs there is a single string instead of a list, the the first string in the list is used:

https://url.spec.whatwg.org/#cannot-be-a-base-url-path-state
> If c is not EOF code point, UTF-8 percent encode c using the simple encode set, and append the result to the first string in url’s path.

This algorithm is only relevant for URLs in the `javascript` scheme.

---

*[components/script/dom/htmlbaseelement.rs, line 87 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyYj0Z8PIEcChaJy-6:-KFyjfl7uGi8bdroRshy:-1788447360) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/script/dom/htmlbaseelement.rs#L87)):*
Changed to `map_or`.

---

*[components/script/dom/urlhelper.rs, line 7 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFy_aEn6wfoiiXjLVSo:-KFyjz5jEtlhKdl0bxsM:710508669) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/script/dom/urlhelper.rs#L7)):*
This was previously `webidl`, but @Ms2ger didn’t like *that* name. It’s quirks because there’s lot of details that don’t make sense to me, and I chose to design the Url inherent methods differently. But we’re still stuck with those as exposed to JS, for compatibility/interop. You can see which functions in https://github.com/servo/rust-url/blob/master/src/quirks.rs are one-liners calling a method, and which have more complex behavior like trimming.

---

*[components/script/dom/urlsearchparams.rs, line 34 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFy_ws1cSIud1lrNRCl:-KFyl2pt0TS7cpm1Zj86:-1933361200) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/script/dom/urlsearchparams.rs#L34)):*
Is that a thing? I can’t find it.

---

*[resources/badcert.html, line 6 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyc35Y9wkvoKDwLBas:-KFylF5xQ4tCG7mDdb0j:1498758871) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/resources/badcert.html#L6)):*
As discussed on IRC, `chrome://foo/bar` has a host `foo` and a path `/bar`, while URLs with `:/` (single slash) after the scheme/protocol don’t have a host.

I don’t think Servo’ chrome URLs need to be compatible with Gecko’s (they weren’t before) but if they do we can have the host be what MDN calls "package name".

---

*[tests/unit/net/chrome_loader.rs, line 10 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFycZi0JFtsCMwEorAO:-KFylpf9TsVd7IWYAECf:-1742223608) ([raw file](https://github.com/servo/servo/blob/0e0e902edd200359db6ae06cd149fb1b0a97fcc6/tests/unit/net/chrome_loader.rs#L10)):*
See previous comment

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9840)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
Some nits... are you planning to squash once we're done reviewing?

---

Reviewed 1 of 53 files at r6, 1 of 1 files at r14, 1 of 1 files at r15.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[components/net/chrome_loader.rs, line 22 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyFGQyGkzKFrsBzSAd:-KFytN3bj56egCw0ZF4_:1097452119) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/net/chrome_loader.rs#L22)):*
Isn't path a `PathBuf` object, not a `URL`? So `path.starts_with(resources)` is operating just on paths, not URLs.

---

*[components/script/script_thread.rs, line 1566 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyVzswrHh0iZEHRWI0:-KFyvKoYu8dU9ysgagY1:-507439142) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/script/script_thread.rs#L1566)):*
Ah yes, javascript URLs don't parse the path silly me.

---

*[components/script/dom/urlhelper.rs, line 7 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFy_aEn6wfoiiXjLVSo:-KFyttAVrmTfdhMW4pQl:-1858702293) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/script/dom/urlhelper.rs#L7)):*
True, but "quirks" is a reserved word in HTML, and this doesn't seem related to quirks mode.

---

*[components/script/dom/urlsearchparams.rs, line 34 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFy_ws1cSIud1lrNRCl:-KFyuhGXfGdtEb11ybr9:-737292594) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/script/dom/urlsearchparams.rs#L34)):*
You're right, I could swear I didn't dream it :)

---

*[resources/badcert.html, line 6 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyc35Y9wkvoKDwLBas:-KFyuyFBX1duM6Z6GY4x:-1621951173) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/resources/badcert.html#L6)):*
I think there's enough docs with `chrome://` URLs that we should try not to be confusing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9840#-:-KFyvSrj6DIU6qO6O9ly:1476298783)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #10706) made this pull request unmergeable. Please resolve the merge conflicts.

SimonSapin committed:
Make chrome: URLs have a "host".

SimonSapin commented:
Not planning to squash each commit is a change that makes sense on its own.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[components/net/chrome_loader.rs, line 22 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyFGQyGkzKFrsBzSAd:-KG-ECNOOYbePLXLBc-B:70624941) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/net/chrome_loader.rs#L22)):*
Yes it is a path. `Path::starts_with` makes sure the file we’re about to open is inside the resources diretory.

---

*[components/script/dom/urlhelper.rs, line 7 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFy_aEn6wfoiiXjLVSo:-KG-EP_n8YBfxnN9pfln:-558421148) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/script/dom/urlhelper.rs#L7)):*
This module is in rust-url, whose 1.0.0 is now published. I’d rather not publish 2.0.0 now just to rename this.

---

*[resources/badcert.html, line 6 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyc35Y9wkvoKDwLBas:-KG-Fjk4GTxRu901UXUp:-1612382764) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/resources/badcert.html#L6)):*
I’ve changed them to `chrome://resources/foo.jpg`, where `resources` is a hard-coded "host".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9840#-:-KG-E67YvX3c7UXjcCbi:-1608054259)*
<!-- Sent from Reviewable.io -->

SimonSapin commented:
Fixed a build error in a Windows-only test and a tidy error.

jdm commented:
It's not clear to me that the chrome URL tests for absolute paths are actually testing absolute paths any more.

asajeffrey commented:
See comments re testing and ensuring that `chrome` URLs stay in the `resources` directory.

---

Reviewed 21 of 21 files at r16.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[components/net/chrome_loader.rs, line 22 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyFGQyGkzKFrsBzSAd:-KG2P_Z1U-RqfnFvKnNO:-396764001) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/net/chrome_loader.rs#L22)):*
Does it? Running https://play.rust-lang.org/?gist=5af410faa55787109fca05d9ff8ed289 shows `/etc/foo/../bar` starts with `/etc/foo`. You probably want `canonicalize`, but that actually touches the file system to follow symlinks, which is probably not what we want.

---

*[components/net/chrome_loader.rs, line 16 \[r16\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG2NCqUlGf4hiOkgePw:-KG2NCqUlGf4hiOkgePx:-305884449) ([raw file](https://github.com/servo/servo/blob/91a1aa763089a8a1cca84946e07fb348651aa3ce/components/net/chrome_loader.rs#L16)):*
Would "localhost" be less magical, at least for file resources? So the URL would be `chrome://localhost/this/that`? Not lovely, but at least `localhost` is a actually a host on most (all?) machines.

---

*[components/net/chrome_loader.rs, line 22 \[r16\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG2OLr5YFdWiH4Llbz2:-KG2OLr5YFdWiH4Llbz3:1402155642) ([raw file](https://github.com/servo/servo/blob/91a1aa763089a8a1cca84946e07fb348651aa3ce/components/net/chrome_loader.rs#L22)):*
This is needed for, e.g. resource files with `#` in their name?

---

*[components/script/dom/urlhelper.rs, line 7 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFy_aEn6wfoiiXjLVSo:-KG2QblkLruEMt7HCsbp:1944744277) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/script/dom/urlhelper.rs#L7)):*
Rats, I didn't spot that. Pity we released `rust-url` as `1.0` rather than `0.9`.

---

*[tests/unit/net/chrome_loader.rs, line 10 \[r16\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG2SGsDVKm371hkvhJx:-KG2SGsDVKm371hkvhJy:-251912750) ([raw file](https://github.com/servo/servo/blob/91a1aa763089a8a1cca84946e07fb348651aa3ce/tests/unit/net/chrome_loader.rs#L10)):*
Can we add a test for `chrome://resources/../README.md` (which should fail). Here `README.md` is a file which is a peer of the resources directory, if we're moving stuff around prior to testing.

---

*[tests/unit/net/chrome_loader.rs, line 16 \[r16\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG2RZ1jsI46mIVB5lcF:-KG2RZ1jsI46mIVB5lcG:349826858) ([raw file](https://github.com/servo/servo/blob/91a1aa763089a8a1cca84946e07fb348651aa3ce/tests/unit/net/chrome_loader.rs#L16)):*
Can we add tests for `chrome://resources/subdir/../badcert.jpg` (which should pass) and `chrome://resources/subdir/../../README.md` (which should fail).

---

*[tests/unit/net/chrome_loader.rs, line 23 \[r16\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG2R3fWKjwPCu_5AD7c:-KG2R3fWKjwPCu_5AD7d:1011299695) ([raw file](https://github.com/servo/servo/blob/91a1aa763089a8a1cca84946e07fb348651aa3ce/tests/unit/net/chrome_loader.rs#L23)):*
Should we also not test `Url::parse("chrome://resources//etc/passwd")`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9840#-:-KG2SziqmnvUtZxhscrH:131686315)*
<!-- Sent from Reviewable.io -->

SimonSapin committed:
More chrome URL hardenning and tests.

SimonSapin commented:
@jdm there is no such thing as a distinction between absolute and non-absolute paths in a parsed URL.

---

Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[components/net/chrome_loader.rs, line 22 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyFGQyGkzKFrsBzSAd:-KG2ghA2YQS4HNAeiP3Y:-1938388920) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/net/chrome_loader.rs#L22)):*
Hmm, yeah, with the code as it is there I think `starts_with` always returns true. I’ll add an explict check for `..` segments.

---

*[components/net/chrome_loader.rs, line 16 \[r16\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG2NCqUlGf4hiOkgePw:-KG2f1Mww0EUrkfqlhny:331881183) ([raw file](https://github.com/servo/servo/blob/91a1aa763089a8a1cca84946e07fb348651aa3ce/components/net/chrome_loader.rs#L16)):*
Disagree. 127.0.0.1 or networking in general are irrelevant to `chrome:` URLs. You linked to https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XUL/Tutorial/The_Chrome_URL , which says:

> The basic syntax of a chrome URL is as follows:
>
>     chrome://<package name>/<part>/<file.xul>

For now `resources` is the only "package" that we have.

---

*[components/net/chrome_loader.rs, line 22 \[r16\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG2OLr5YFdWiH4Llbz2:-KG2fdHm9xfUWF-puiqN:335769065) ([raw file](https://github.com/servo/servo/blob/91a1aa763089a8a1cca84946e07fb348651aa3ce/components/net/chrome_loader.rs#L22)):*
The URL parser percent-encodes some characters (including some ASCII punctuation and anything non-ASCII). Maybe it doesn’t matter as long as nothing in the `resources` directory has a name with these characters, but the correct thing to do is still to percent-decode path segments before using them.

---

*[tests/unit/net/chrome_loader.rs, line 10 \[r16\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG2SGsDVKm371hkvhJx:-KG2oT8LOvJ2rT4mODWP:-499146009) ([raw file](https://github.com/servo/servo/blob/91a1aa763089a8a1cca84946e07fb348651aa3ce/tests/unit/net/chrome_loader.rs#L10)):*
Done.

---

*[tests/unit/net/chrome_loader.rs, line 16 \[r16\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG2RZ1jsI46mIVB5lcF:-KG2oTt3JOE5wZd2eEUi:17141493) ([raw file](https://github.com/servo/servo/blob/91a1aa763089a8a1cca84946e07fb348651aa3ce/tests/unit/net/chrome_loader.rs#L16)):*
Done, though both URL strings parse as `chrome://resources/badcert.jpg` and therefore pass. (The URL parser ignores/removes `..` path segments that would make the path go "beyond" the root.)

---

*[tests/unit/net/chrome_loader.rs, line 23 \[r16\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG2R3fWKjwPCu_5AD7c:-KG2olZTSRVZEWViHXoi:527908498) ([raw file](https://github.com/servo/servo/blob/91a1aa763089a8a1cca84946e07fb348651aa3ce/tests/unit/net/chrome_loader.rs#L23)):*
Done, though this is really testing the URL parser more than the chrome loader.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9840#-:-KG2elZbj-EJQ3cNOxMm:-1614266856)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
We can pass in percent-encoded special chars like `/` and `..` though.

---

Reviewed 2 of 2 files at r17.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[components/net/chrome_loader.rs, line 22 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyFGQyGkzKFrsBzSAd:-KG2t4Xl7mVccWdZp55B:127724279) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/net/chrome_loader.rs#L22)):*
According to https://play.rust-lang.org/?gist=8a2d1a4e1a2d6f9d9c9e3e799aa88777, `/etc/foo/../bar` doesn't start with `/etc/bar`. I'm concerned that if we just special-case `..` we'll miss some case which allows access to the whole file system. I still think `canonicalize` is our best bet.

---

*[tests/unit/net/chrome_loader.rs, line 16 \[r16\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG2RZ1jsI46mIVB5lcF:-KG2tuS4rgOQSzCMNdYG:685802953) ([raw file](https://github.com/servo/servo/blob/91a1aa763089a8a1cca84946e07fb348651aa3ce/tests/unit/net/chrome_loader.rs#L16)):*
Can we percent-encode the dots to get them to the file system?

---

*[tests/unit/net/chrome_loader.rs, line 23 \[r16\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG2R3fWKjwPCu_5AD7c:-KG2u4I5XTLteTXIt00K:-1967028043) ([raw file](https://github.com/servo/servo/blob/91a1aa763089a8a1cca84946e07fb348651aa3ce/tests/unit/net/chrome_loader.rs#L23)):*
Can we percent-encode `/etc/passwd` then?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9840#-:-KG2uF6yqIRlazNg3Ker:1761917447)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #10823) made this pull request unmergeable. Please resolve the merge conflicts.

SimonSapin commented:
Review status: 72 of 78 files reviewed at latest revision, 7 unresolved discussions.

---

*[components/net/chrome_loader.rs, line 22 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyFGQyGkzKFrsBzSAd:-KG2vqhs13tm6scwfuZ0:-202863966) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/net/chrome_loader.rs#L22)):*
Since `canonicalize` resolves symlinks, using it with `starts_width` won’t work in the path to the resources contains a symlink.

---

*[tests/unit/net/chrome_loader.rs, line 16 \[r16\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG2RZ1jsI46mIVB5lcF:-KG2vg6QJ-nIszelXocg:-459602462) ([raw file](https://github.com/servo/servo/blob/91a1aa763089a8a1cca84946e07fb348651aa3ce/tests/unit/net/chrome_loader.rs#L16)):*
Added a test with chrome://resources/%2e%2e/README.md

---

*[tests/unit/net/chrome_loader.rs, line 23 \[r16\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG2R3fWKjwPCu_5AD7c:-KG2vm8s76CTamGpuJCI:-1694351514) ([raw file](https://github.com/servo/servo/blob/91a1aa763089a8a1cca84946e07fb348651aa3ce/tests/unit/net/chrome_loader.rs#L23)):*
Added a test for that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9840)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
Reviewed 6 of 6 files at r18.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[components/net/chrome_loader.rs, line 22 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyFGQyGkzKFrsBzSAd:-KG3-AmXhHcaS15pZfqj:1371389493) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/net/chrome_loader.rs#L22)):*
Canonicalize the resources path?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9840)*
<!-- Sent from Reviewable.io -->

SimonSapin commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[components/net/chrome_loader.rs, line 22 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9840#-KFyFGQyGkzKFrsBzSAd:-KG36SzDaKe_Od_iNfo9:-499146009) ([raw file](https://github.com/servo/servo/blob/f0b9d05c731db6b953274fb8ee8a07fcd3474205/components/net/chrome_loader.rs#L22)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9840)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
Getting there...

---

Reviewed 1 of 1 files at r19.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[components/net/chrome_loader.rs, line 21 \[r19\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG36lzA-AmTLMAKwsuA:-KG36lzB-Irr8tw7z1zc:940177819) ([raw file](https://github.com/servo/servo/blob/9762e39d6d32fb89af2be49736023b90eb65fcbe/components/net/chrome_loader.rs#L21)):*
Can we use `try!` for this? (May need a more precise error type).

---

*[components/net/chrome_loader.rs, line 27 \[r19\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG37Inv61O15Dd37i7E:-KG37Inv61O15Dd37i7F:1816442704) ([raw file](https://github.com/servo/servo/blob/9762e39d6d32fb89af2be49736023b90eb65fcbe/components/net/chrome_loader.rs#L27)):*
We don't need to check for bad segments if we check at the end that the result is inside the resources dir.

---

*[components/net/chrome_loader.rs, line 31 \[r19\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG375z73JMjmZvVnoKM:-KG375z73JMjmZvVnoKN:1700997001) ([raw file](https://github.com/servo/servo/blob/9762e39d6d32fb89af2be49736023b90eb65fcbe/components/net/chrome_loader.rs#L31)):*
Ditto.

---

*[components/net/chrome_loader.rs, line 32 \[r19\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG37TtVaQjd1n5VD7Zp:-KG37TtVaQjd1n5VD7Zq:-427016147) ([raw file](https://github.com/servo/servo/blob/9762e39d6d32fb89af2be49736023b90eb65fcbe/components/net/chrome_loader.rs#L32)):*
Add a check that `path.starts_with(resources)`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9840#-:-KG37d89Fa29hXU_Bjxs:671491595)*
<!-- Sent from Reviewable.io -->

SimonSapin commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[components/net/chrome_loader.rs, line 21 \[r19\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG36lzA-AmTLMAKwsuA:-KG38HrS3EqER8KbR7-u:1970699360) ([raw file](https://github.com/servo/servo/blob/9762e39d6d32fb89af2be49736023b90eb65fcbe/components/net/chrome_loader.rs#L21)):*
Something is very wrong in the configuration of the resource directory if this one call fails, so I think we should panic in that case. It’s not about the `url` argument is valid.

---

*[components/net/chrome_loader.rs, line 27 \[r19\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG37Inv61O15Dd37i7E:-KG38o1ZCIzGQ2sJ3pK8:1740281843) ([raw file](https://github.com/servo/servo/blob/9762e39d6d32fb89af2be49736023b90eb65fcbe/components/net/chrome_loader.rs#L27)):*
Removed.

---

*[components/net/chrome_loader.rs, line 31 \[r19\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG375z73JMjmZvVnoKM:-KG38ovW3r0zVcu4eVGx:-2035699351) ([raw file](https://github.com/servo/servo/blob/9762e39d6d32fb89af2be49736023b90eb65fcbe/components/net/chrome_loader.rs#L31)):*
I think this is better. The error types are not compatible, so using `try!` would require `.map_err(|_| ())`, and the match syntax allows us to have the `if` guard in the same construct.

---

*[components/net/chrome_loader.rs, line 32 \[r19\]](https://reviewable.io:443/reviews/servo/servo/9840#-KG37TtVaQjd1n5VD7Zp:-KG393lD1clPT__ws8AK:-499146009) ([raw file](https://github.com/servo/servo/blob/9762e39d6d32fb89af2be49736023b90eb65fcbe/components/net/chrome_loader.rs#L32)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9840)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
Reviewed 1 of 1 files at r20.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9840)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
OK, at this point I'm done. I'd rather get a fresh pair of eyes to look over it, but if you reckon it's urgent you can r=me.

SimonSapin commented:
@bors-servo r=asajeffrey

bors-servo commented:
:pushpin: Commit 45acaea has been approved by `asajeffrey`

<!-- @bors-servo r=asajeffrey 45acaeafea850f5ed97871003e6e71fe447e2cd8 -->

bors-servo commented:
:hourglass: Testing commit 45acaea with merge 882ce08...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/534)

SimonSapin commented:
Updated lockfile.

@bors-servo r=asajeffrey

bors-servo commented:
:pushpin: Commit e662605 has been approved by `asajeffrey`

<!-- @bors-servo r=asajeffrey e66260513834e4ffafcd1539649432068c0ecc70 -->

bors-servo commented:
:hourglass: Testing commit e662605 with merge 84ab7e9...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/493), [arm32](http://build.servo.org/builders/arm32/builds/526), [arm64](http://build.servo.org/builders/arm64/builds/511), [linux-dev](http://build.servo.org/builders/linux-dev/builds/520), [linux-rel](http://build.servo.org/builders/linux-rel/builds/571), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/535), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/521), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/548)
